store/postgres/src/store.rs

```diff
@@ -23,18 +24,134 @@ use graph::{tokio, tokio::timer::Interval};
+#[derive(Clone, Debug)]
+enum Subscription {
+    // We have entities in both kinds of subscriptions; since we clone the
+    // entire subscription in a few places, we make the entities an Arc
+    // to make the cloning cheaper.
+    Changes {
+        entities: Arc<Vec<SubgraphEntityPair>>,
+        sender: Sender<EntityChange>,
+    },
+    Ticks {
+        entities: Arc<Vec<SubgraphEntityPair>>,
+        sender: Sender<BlockTick>,
+    },
+}
+
+/// Track which subscriptions have had changes to their blocks. The entries
+/// in the set are the id's of subscriptions for which we've seen an
+/// EntityChange matching their interesting entities. When the corresponding
+/// BlockTick comes by, we pass a BlockTick to the receiver and remove
+/// the id from the set
+#[derive(Clone)]
+struct DirtyBlocks {
+    ids: Arc<RwLock<std::collections::HashSet<String>>>,
+}
+
+impl DirtyBlocks {
+    fn new() -> Self {
+        DirtyBlocks {
+            ids: Arc::new(RwLock::new(std::collections::HashSet::new())),
+        }
+    }
+
+    fn remove(&self, id: &str) -> bool {
+        self.ids.write().unwrap().remove(id)
+    }
+
+    fn insert(&self, id: String) -> bool {
+        self.ids.write().unwrap().insert(id)
+    }
+
+    fn contains(&self, id: &str) -> bool {
+        self.ids.read().unwrap().contains(id)
+    }
+}
+
+/// Subscription translates from the internal `StoreEvent` to either
+/// `EntityChange` or `BlockTick`, depending on what the subscriber is
+/// interested in. The 'other' kind of event for a subscriber is simply
+/// dropped on the floor
+///
+/// This is not a full implementation of a `futures::sink::Sink` since
+/// we don't need that right now, but could be expanded to that
+impl Subscription {
+    fn send(
+        self,
+        e: StoreEvent,
+        id: &String,
+        dirty_blocks: &DirtyBlocks,
+    ) -> impl Future<Item = (), Error = ()> {
+        match (self, e) {
+            (Subscription::Changes { sender, .. }, StoreEvent::Change(change)) => {
+                let s = sender.send(change);
+                let unit = s.map(|_| ()).map_err(|_| ());
+                Either::A(Either::A(unit))
+            }
+            (Subscription::Ticks { sender, .. }, StoreEvent::Tick(tick)) => {
+                let dirty = dirty_blocks.remove(id);
+                if dirty {
+                    let s = sender.send(tick);
+                    let unit = s.map(|_| ()).map_err(|_| ());
+                    Either::A(Either::B(unit))
+                } else {
+                    Either::B(ok(()))
+                }
+            }
+            _ => Either::B(ok(())),
+        }
+    }
+
+    fn poll_ready(&mut self) -> Poll<(), SendError<()>> {
+        match self {
+            Subscription::Changes { sender, .. } => sender.poll_ready(),
+            Subscription::Ticks { sender, .. } => sender.poll_ready(),
+        }
+    }
+
+    fn matches(&self, event: &StoreEvent, id: &String, dirty_blocks: &DirtyBlocks) -> bool {
```

Comment:
The semantics of this method are not entirely clear to me. Could you add a comment here, including a description of why this is needed in addition to the other matching method below?